### PR TITLE
[TEP-0076] Add indexing into array for taskrun params reference

### PR DESCRIPTION
**Pull request by Yongxuanzhang**
# Changes
This commit provides the indexing into array for params and gated by
alpha feature flag. Before this commit we can only refer to the whole
array for taskrun params, with this feature we can refer to array's element such as
$(params.param-name[0]).

/kind feature

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [x] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
Indexing into array for taskrun params is now an alpha feature, element of array params can be accessed via $(params.param-name[i]).
```

### CODE REVIEW

File: pkg/reconciler/taskrun/validate_resources.go [hunk 1]
```diff
@@ -353,3 +369,200 @@ func missingKeysofObjectResults(tr *v1beta1.TaskRun, specResults []v1beta1.TaskR
 	}
 	return findMissingKeys(neededKeys, providedKeys)
 }
+
+func validateParamArrayIndex(ctx context.Context, params []v1beta1.Param, spec *v1beta1.TaskSpec) error {
```
Comment (lbernick) [inline pkg/reconciler/taskrun/validate_resources.go:373]: this block of code looks extremely similar to what was introduced in the pipelinerun reconciler. Can any of it be moved to a common location?

File: pkg/reconciler/taskrun/resources/apply_test.go [hunk 2]
```diff
@@ -697,7 +871,7 @@ func TestApplyArrayParameters(t *testing.T) {
 	}}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			got := resources.ApplyParameters(tt.args.ts, tt.args.tr, tt.args.dp...)
+			got := resources.ApplyParameters(context.Background(), tt.args.ts, tt.args.tr, tt.args.dp...)
```
[the remark this answers is not in the document]
Comment (Yongxuanzhang) [reply pkg/reconciler/taskrun/resources/apply_test.go:874]: For param indexing they will remain the reference string. e.g. "$(params.array-params[1])".
Should I write test for this behaviour?

**Comment by Yongxuanzhang** [comment]
@ywluogg

**Comment by Yongxuanzhang** [comment]
#4723

File: pkg/reconciler/taskrun/validate_resources.go (hunk 1 above)
Comment (lbernick) [reply pkg/reconciler/taskrun/validate_resources.go:359]: I'm not sure if you're referring to `extractParamIndex` here or `extractParamIndexes` in the PR reconciler but either way yes they could be pulled out. I'd create a new package `pkg/params` and put it there in a separate PR

File: pkg/reconciler/taskrun/resources/apply_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (dibyom) [reply pkg/reconciler/taskrun/resources/apply_test.go:874]: I think what @lbernick is referring to is that by passing `context.Background()`,  the Alpha API fields will not be set (`config.FromContextOrDefaults(ctx)` will return the default stable value). So we'd have to pass a context that has alpha feature flag set: https://github.com/tektoncd/pipeline/tree/main/docs/developers#unit-testing-with-feature-gates

[the remark this answers is not in the document]
Comment (lbernick) [reply pkg/reconciler/taskrun/resources/apply_test.go:874]: ah I see this makes sense now

**Review by lbernick — APPROVED** [review]

**Pull request merged**